File: tests/performance/drivers/jdbc/app/src/net/snowflake/perf/ConcurrentExecution.java
package net.snowflake.perf;

import java.sql.Connection;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Properties;
import java.util.concurrent.BrokenBarrierException;
import java.util.concurrent.CyclicBarrier;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

/**
 * Concurrent SELECT bursts: one JDBC connection per worker. Connections are opened and set up
 * before burst timing; wall time covers query and fetch only.
 */
final class ConcurrentExecution {

  private static final int BARRIER_TIMEOUT_S = 120;

  private ConcurrentExecution() {}

  static QueryExecution.FetchTestOutput execute(
      String driverType,
      String url,
      Properties props,
      List<String> setupQueries,
      String sql,
      int warmupIterations,
      int iterations,
      int workerCount)
      throws Exception {
    System.out.println("\n=== Executing Concurrent SELECT Test ===");
    System.out.println("Query: " + sql);
    System.out.println(
        "Workers: " + workerCount + " connections (one statement per connection)");

    System.out.println(
        "Opening " + workerCount + " worker connections (excluded from burst timing)...");
    ExecutorService pool = Executors.newFixedThreadPool(workerCount);
    try {
      try (WorkerConnections workers =
          openWorkerConnections(pool, driverType, url, props, setupQueries, workerCount)) {
        System.out.println("Worker connections ready");

        for (int i = 1; i <= warmupIterations; i++) {
          System.out.println("  Warmup burst " + i + "/" + warmupIterations);
          runBurst(pool, workers.connections, sql);
        }

        ResourceMonitor monitor = new ResourceMonitor(100);
        monitor.start();
        List<QueryExecution.IterationResult> results = new ArrayList<>();
        for (int i = 1; i <= iterations; i++) {
          QueryExecution.IterationResult result = runBurst(pool, workers.connections, sql);
          results.add(result);
          System.out.printf(
              "  Iteration %d/%d: burst=%.3fs  throughput=%.0f rows/s  rows=%d%n",
              i, iterations, result.queryTimeS, result.throughputRowsS, result.rowCount);
        }
        List<ResourceMonitor.Sample> timeline = monitor.stop();

        validateRowCounts(results, workerCount);
        printStatistics(results);
        System.out.println("  Memory timeline: " + timeline.size() + " samples collected");
        return new QueryExecution.FetchTestOutput(results, timeline);
      }
    } finally {
      pool.shutdownNow();
      try {
        pool.awaitTermination(30, TimeUnit.SECONDS);
      } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
      }
    }
  }

  private static WorkerConnections openWorkerConnections(
      ExecutorService pool,
      String driverType,
      String url,
      Properties props,
      List<String> setupQueries,
      int workerCount)
      throws Exception {
    WorkerConnections workers = new WorkerConnections();
    if (!setupQueries.isEmpty()) {
      System.out.println(
          "Running setup queries on " + workerCount + " worker connections...");
    }
    List<Future<Connection>> futures = new ArrayList<>(workerCount);
    for (int i = 0; i < workerCount; i++) {
      futures.add(
          pool.submit(
              () -> {
                Connection conn = ConnectionFactory.connect(driverType, url, props);
                ConnectionFactory.executeSetupQueries(conn, setupQueries, false);
                return conn;
              }));
    }

    Exception firstError = null;
    for (Future<Connection> future : futures) {
      try {
        workers.connections.add(future.get());
      } catch (ExecutionException e) {
        Exception workerError = unwrap(e);
        if (firstError == null) {
          firstError = workerError;
        } else {
          System.err.println("Additional worker connection failure:");
          workerError.printStackTrace(System.err);
        }
      } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
        workers.close();
        throw new IllegalStateException("Worker connection setup interrupted", e);
      }
    }
    if (firstError != null) {
      workers.close();
      throw firstError;
    }
    return workers;
  }

  private static QueryExecution.IterationResult runBurst(
      ExecutorService pool, List<Connection> connections, String sql) throws Exception {
    int workerCount = connections.size();
    long[] burstStartNs = {0L};
    CyclicBarrier barrier =
        new CyclicBarrier(workerCount + 1, () -> burstStartNs[0] = System.nanoTime());

    List<Future<Long>> futures = new ArrayList<>(workerCount);
    for (Connection conn : connections) {
      futures.add(
          pool.submit(
              () -> {
                barrier.await(BARRIER_TIMEOUT_S, TimeUnit.SECONDS);
                return QueryExecution.fetchAllRows(conn, sql);
              }));
    }
    try {
      barrier.await(BARRIER_TIMEOUT_S, TimeUnit.SECONDS);
    } catch (TimeoutException e) {
      throw new IllegalStateException("Concurrent burst barrier timed out", e);
    } catch (BrokenBarrierException e) {
      throw new IllegalStateException("Concurrent burst barrier broken", e);
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      throw new IllegalStateException("Concurrent burst barrier interrupted", e);
    }
    double cpuStart = Common.processCpuSeconds();

    List<Long> workerRows = new ArrayList<>(workerCount);
    Exception firstError = null;
    for (Future<Long> future : futures) {
      try {
        workerRows.add(future.get());
      } catch (ExecutionException e) {
        Exception workerError = unwrap(e);
        if (firstError == null) {
          firstError = workerError;
        } else {
          System.err.println("Additional worker failure:");
          workerError.printStackTrace(System.err);
        }
      }
    }
    if (firstError != null) {
      throw firstError;
    }

    double burstWallS = (System.nanoTime() - burstStartNs[0]) / 1_000_000_000.0;
    long perWorkerRows = workerRows.get(0);
    for (int i = 1; i < workerRows.size(); i++) {
      if (!workerRows.get(i).equals(perWorkerRows)) {
        throw new IllegalStateException(
            "Workers returned unequal row counts: " + workerRows);
      }
    }
    long totalRows = perWorkerRows * workerCount;
    double cpuTimeS = Common.processCpuSeconds() - cpuStart;
    double throughput = burstWallS > 0.0 ? totalRows / burstWallS : 0.0;
    return new QueryExecution.IterationResult(
        System.currentTimeMillis(),
        burstWallS,
        burstWallS,
        totalRows,
        cpuTimeS,
        Common.peakRssMb(),
        workerCount,
        throughput);
  }

  private static Exception unwrap(ExecutionException e) {
    Throwable cause = e.getCause();
    if (cause instanceof Exception) {
      return (Exception) cause;
    }
    return e;
  }

  private static void validateRowCounts(
      List<QueryExecution.IterationResult> results, int workerCount) {
    if (results.isEmpty()) {
      return;
    }
    long expected = results.get(0).rowCount;
    if (expected == 0) {
      throw new IllegalStateException(
          "Row count baseline is 0 — refusing to use 0 as a concurrent-burst baseline.");
    }
    long perWorker = expected / workerCount;
    for (int i = 0; i < results.size(); i++) {
      long actual = results.get(i).rowCount;
      if (actual != expected) {
        throw new IllegalStateException(
            String.format(
                "Row count mismatch: iteration %d returned %d rows, expected %d (%d workers × %d)",
                i, actual, expected, workerCount, perWorker));
      }
    }
    System.out.printf(
        "All %d bursts returned %d rows (%d × %d)%n",
        results.size(), expected, workerCount, perWorker);
  }

  private static void printStatistics(List<QueryExecution.IterationResult> results) {
    if (results.isEmpty()) {
      return;
    }
    List<Double> burstTimes = new ArrayList<>(results.size());
    List<Double> throughputs = new ArrayList<>(results.size());
    for (QueryExecution.IterationResult result : results) {
      burstTimes.add(result.queryTimeS);
      throughputs.add(result.throughputRowsS);
    }
    System.out.println("\nSummary:");
    System.out.printf(
        "  Burst wall: median=%.3fs  min=%.3f  max=%.3f%n",
        median(burstTimes), Collections.min(burstTimes), Collections.max(burstTimes));
    System.out.printf(
        "  Throughput: median=%.0f rows/s  min=%.0f  max=%.0f%n",
        median(throughputs), Collections.min(throughputs), Collections.max(throughputs));
  }

  private static double median(List<Double> values) {
    List<Double> sorted = new ArrayList<>(values);
    Collections.sort(sorted);
    int n = sorted.size();
    if (n % 2 == 1) {
      return sorted.get(n / 2);
    }
    return (sorted.get(n / 2 - 1) + sorted.get(n / 2)) / 2.0;
  }

  private static final class WorkerConnections implements AutoCloseable {
    final List<Connection> connections = new ArrayList<>();

    @Override
    public void close() {
      for (int i = connections.size() - 1; i >= 0; i--) {
        try {
          connections.get(i).close();
        } catch (SQLException ignored) {
          // ignore
        }
      }
    }
  }
}

File: tests/performance/drivers/python/app/concurrent_execution.py
"""Same-connection concurrent SELECT bursts.

One Snowflake connection, N worker threads, each with its own cursor.
"""

from concurrent.futures import ThreadPoolExecutor, as_completed
import statistics
import threading
import time

from common import get_peak_rss_mb, print_timing_stats
from query_execution import _FETCH_STRATEGIES
from resource_monitor import ResourceMonitor


def execute_concurrent_test(
    conn,
    sql_command,
    warmup_iterations,
    iterations,
    worker_count,
    fetch_mode="fetchmany",
):
    """Run warmup + measured bursts. Returns (results, memory_timeline)."""
    print("\n=== Executing Concurrent SELECT Test ===")
    print(f"Query: {sql_command}")
    print(f"Workers: {worker_count} threads on one connection")
    print(f"Fetch mode: {fetch_mode}")

    fetch_fn = _FETCH_STRATEGIES[fetch_mode]

    for i in range(warmup_iterations):
        print(f"  Warmup burst {i + 1}/{warmup_iterations}")
        _run_burst(conn, sql_command, worker_count, fetch_fn)

    monitor = ResourceMonitor(interval_s=0.1)
    monitor.start()

    results = []
    for i in range(iterations):
        results.append(_run_burst(conn, sql_command, worker_count, fetch_fn))
        print(
            f"  Iteration {i + 1}/{iterations}: "
            f"burst={results[-1]['query_time_s']:.3f}s  "
            f"throughput={results[-1]['throughput_rows_s']:.0f} rows/s  "
            f"rows={results[-1]['row_count']}"
        )

    memory_timeline = monitor.stop()

    _validate_row_counts(results, worker_count)
    _print_statistics(results)
    print(f"  Memory timeline: {len(memory_timeline)} samples collected")

    return results, memory_timeline


def _barrier_timeout_s(worker_count: int) -> float:
    return min(900.0, max(120.0, worker_count * 0.5))


def _run_burst(conn, sql, worker_count, fetch_fn):
    burst_start = {"t": None}
    barrier_timeout = _barrier_timeout_s(worker_count)

    def mark_start():
        burst_start["t"] = time.perf_counter()

    barrier = threading.Barrier(worker_count + 1, action=mark_start)
    cpu_start = time.process_time()

    def worker():
        barrier.wait(timeout=barrier_timeout)
        with conn.cursor() as cursor:
            cursor.execute(sql)
            return fetch_fn(cursor)

    with ThreadPoolExecutor(max_workers=worker_count) as pool:
        futures = [pool.submit(worker) for _ in range(worker_count)]
        barrier.wait(timeout=barrier_timeout)
        worker_results = []
        for future in as_completed(futures):
            worker_results.append(future.result())

    burst_wall_s = time.perf_counter() - burst_start["t"]
    cpu_time_s = time.process_time() - cpu_start
    total_rows = sum(worker_results)
    per_worker_rows = set(worker_results)
    if len(per_worker_rows) != 1:
        raise RuntimeError(
            f"Workers returned unequal row counts: {sorted(per_worker_rows)}"
        )

    return {
        "timestamp": int(time.time() * 1000),
        "query_time_s": burst_wall_s,
        "fetch_time_s": burst_wall_s,
        "row_count": total_rows,
        "cpu_time_s": cpu_time_s,
        "peak_rss_mb": get_peak_rss_mb(),
        "worker_count": worker_count,
        "throughput_rows_s": total_rows / burst_wall_s if burst_wall_s > 0 else 0.0,
    }


def _validate_row_counts(results, worker_count):
    if not results:
        return
    expected = results[0]["row_count"]
    if expected == 0:
        raise RuntimeError(
            "Row count baseline is 0 — refusing to use 0 as a concurrent-burst baseline."
        )
    if expected % worker_count != 0:
        raise RuntimeError(
            f"Total row count {expected} is not divisible by worker_count {worker_count}"
        )
    per_worker = expected // worker_count
    for i, result in enumerate(results):
        if result["row_count"] != expected:
            raise RuntimeError(
                f"Row count mismatch: iteration {i} returned {result['row_count']} rows, "
                f"expected {expected} ({worker_count} workers × {per_worker})"
            )
    print(
        f"✓ All {len(results)} bursts returned {expected} rows "
        f"({worker_count} × {per_worker})"
    )


def _print_statistics(results):
    burst_times = [r["query_time_s"] for r in results]
    throughputs = [r["throughput_rows_s"] for r in results]
    print("\nSummary:")
    print_timing_stats("Burst wall", burst_times)
    print(
        f"  Throughput: median={statistics.median(throughputs):.0f} rows/s  "
        f"min={min(throughputs):.0f}  max={max(throughputs):.0f}"
    )
